### PR TITLE
PT-1320 spec for updated data types

### PR DESCRIPTION
**Pull request by franck-fourel**
documentation/specification only

### CODE REVIEW

File: docs/UpdatedData.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+{
+  "time": "2020-05-12T08:50:08.000Z",
+  "timezoneOffset": 120,
+  "deviceTime": "2020-05-12T08:50:08",
+  "deviceId": "IdOfTheDevice",
+  "type": "bolus",
+  "subType": "normal",
+  "normal": 3.5,
+  "expectedNormal": 4.0, 
+  "prescriptor": "system-altered"
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: Is it another prescriptor type undocumented or the example value should be "auto-altered" ?

File: docs/UpdatedData.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+- net: total amount of recommended insulin
+
+Here is an example of what can be sent with the related meaning:
+- `recommended.net` is the system recommendation
+- `bolus.normal` is the value delivered by the insulin pump.
+- `bolus.expectedNormal` is the original value that has been requested to the insulin pump.
+- `bolus.prescriptor` is a new field that describes the origin of the bolus. Details are defined in the below food section. 
+
+And the additional field we would need:
+- `entryTime` is a UTC string timestamp that defines at what time the patient has entered the meal. This field is optional. It takes the same format as `time` field.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: Can't we use the createdTime field ? What does represent time field in that case ?

File: docs/UpdatedData.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+3 types of bolus events are available as of now in the system:
+- normal
+- square
+- dual/square
+
+Here we are introducing 2 new fields in the bolus objects:
+- `prescriptor`: same as above in `food`. This field is optional. 
+- `insulinOnBoard`: amount of active insulin estimated by the system. This field will be accepted when `prescriptor` is either `auto` or `auto-altered`. It will be ignored for `manual` entries.
+
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: This field already exist for wizard objects. should it have same rules of validation (range) ?

File: docs/UpdatedData.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  "normal": 3.5,
+  "expectedNormal": 4.0, 
+  "prescriptor": "system-altered"
+}
+```
+
+## physical activity
+
+We need additional fields to get the time at which the physical activity is created, and the last time it was updated by the patient:
+- `entryTime` is a UTC string timestamp that defines at what time the patient has entered the physical activity. This field is optional. It takes the same format as `time` field.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: Can't we use the createdTime field ? What does represent time field in that case ?

File: docs/UpdatedData.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+- `lastUpdatedTime` is a UTC string timestamp that gives the last time the patient has updated the physical activity. This field is optional. It takes the same format as `time` field.
+  - `lastUpdatedTime` >= `entryTime`
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: Does this also means that terminal will have to update the data ? 
Couldn't we use an eventType  field with values 'start'/'update' (similar to tidepool [devicesensor](https://github.com/tidepool-org/platform/blob/master/data/types/device/devicesensor/devicesensor.go) events) and store two objects ?

File: docs/UpdatedData.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+}
+```
+
+## Alarm events
+Leveraging the `deviceEvent` type with the already defined `alarm` subType. We add couple of fields to get more details on alarms and acknowledgement. 
+
+- `alarmLevel`: `alarm | alert` 
+- `alarmCode`: code of the alarm. This field is optional. 
+- `alarmLabel`: label or description of the alarm. This field is optional. 
+- `acknowledgeTime`: is a UTC string timestamp that indicates at what time the patient has acknowledged the alam.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: Same remark than the one the "lastUpdatedTime" field in physical activity. 
We could store two object the first one with eventType="start" and a second one with the eventType="acknowledge", for example.

File: docs/UpdatedData.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+  "timezoneOffset": 60
+}
+```
+
+
+## Zen mode
+Leveraging the `deviceEvent` type and creating a new `zen` subType.
+
+- `subType`: `zen`
+- `endTime`: is a UTC string timestamp that indicates at what time the event is terminated.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: Same remark than the one the "lastUpdatedTime" field in physical activity.
We could store two object the first one with eventType="start" and a second one with the eventType="end", for example.

File: docs/UpdatedData.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+  "time": "2020-05-12T08:50:08.000Z",
+  "timezoneOffset": 60
+}
+```
+
+## Confidential mode
+Leveraging the `deviceEvent` type and creating a new `confidential` subType.
+
+- `subType`: `confidential`
+- `endTime`: is a UTC string timestamp that indicates at what time the event is terminated.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: Same remark than the one the "lastUpdatedTime" field in physical activity.
We could store two object the first one with eventType="start" and a second one with the eventType="end", for example.

**Review by CedricPerotto — COMMENTED** [review]
Some questions and remarks...

File: docs/UpdatedData.md (hunk 2 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: the time is the effective time of the meal. You took your meal at noon but you entered it in the system 10 minutes before; at 11:50am. In that case, entryTime will be 11:50am while time will be set to noon.

File: docs/UpdatedData.md (hunk 1 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: it's a typo. 
To be honest I don't like at all the `system-altered` neither the `auto-altered`. 
Any proposal is welcome ;-)

File: docs/UpdatedData.md (hunk 3 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: Yes, the range values will be the same as the one defined in wizard.

File: docs/UpdatedData.md (hunk 4 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: same as above with bolus object

File: docs/UpdatedData.md (hunk 5 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: That's a good point. 
We should avoid data update. 
My main concern is that the physical activity object can be updated on the handset during 24 hours. It's ok to have several objects to reflect those updates but It means that we have to be able to link the different objects. Adding a key at the input object level would ease the link. 
Something we should discuss.

File: docs/UpdatedData.md (hunk 7 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: same as above.

File: docs/UpdatedData.md (hunk 8 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: same as above

File: docs/UpdatedData.md (hunk 6 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: let me try to propose something ...

File: docs/UpdatedData.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+- net: total amount of recommended insulin
+
+Here is an example of what can be sent with the related meaning:
+- `recommended.net` is the system recommendation
+- `bolus.normal` is the value delivered by the insulin pump.
+- `bolus.expectedNormal` is the original value that has been requested to the insulin pump.
+- `bolus.prescriptor` is a new field that describes the origin of the bolus. Details are defined in the below food section. 
+
+And the additional field we would need:
+- `entryTime` is a UTC string timestamp that defines at what time the patient has entered the meal. This field is optional. It takes the same format as `time` field.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: maybe inputTime instead of entryTime ?

File: docs/UpdatedData.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+  "deviceTime": "2020-05-12T12:00:08",
+  "deviceId": "IdOfTheDevice",
+  "type": "bolus",
+  "subType": "biphasic",
+  "bolusId": "Bo123456789",
+  "part": 1,
+  "normal": 3.5,
+  "expectedNormal": 4.0, 
+  "linkedBolus": {
+    "Normal": 3.5,
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]:
```suggestion
    "normal": 3.5,
```

File: docs/UpdatedData.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+    "Duration": 3600000
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]:
```suggestion
    "duration": 3600000
```
Here we express duration in ms ? but for physical activity we have it in a data structure with value and units, I think we should have only one way to represent duration.

File: docs/UpdatedData.md [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+  "part": 2,
+  "normal": 3.5,
+  "prescriptor": "system"
+}
+```
+
+## physical activity
+
+We need additional fields to get the time at which the physical activity is created, and the last time it was updated by the patient:
+- `entryTime` is a UTC string timestamp that defines at what time the patient has entered the physical activity. This field is optional. It takes the same format as `time` field.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: maybe inputTime instead of entryTime ?

File: docs/UpdatedData.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+- `eventType`: type of event, either `start` or `end`
+  - `start` defines the beginning of the event. The `duration` is the estimated one. The `time` field gives the actual start time of the event.
+  - `stop` gives the end of the event. The `duration` is the actual duration. The `time` field gives the actual end time of the event. 
+- `eventID`: unique ID provided by the client that is used to link stop and start events.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]:
```suggestion
- `eventId`: unique ID provided by the client that is used to link stop and start events.
```

File: docs/UpdatedData.md [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+  "timezoneOffset": 60
+}
+```
+
+## Zen mode
+Leveraging the `deviceEvent` type and creating a new `zen` subType.
+
+- `subType`: `zen`
+- `duration`: is a structured object that gives the duration of the event. 
+- `eventType`: `start | stop` is the type of event for the given event.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md:251]: Don't we also need eventId field ?

File: docs/UpdatedData.md [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+  - `start`: event created by the system. The `duration` attached to this object is the expected duration of the event.
+  - `stop`: the event is stopped. The `duration` attached to this object is the actual duration of the event.
+
+```json
+{
+  "type": "deviceEvent",
+  "subType": "confidential",
+  "eventType": "start",
```
Comment (CedricPerotto) [inline docs/UpdatedData.md:274]: Don't we also need eventId field ?

File: docs/UpdatedData.md [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+  "expectedNormal": 4.0, 
+  "prescriptor": "hybrid"
+}
+```
+
+## biphasic bolus
+
+A `biphasic` bolus is a 2 parts bolus that is defined by the system. Below is the definition for this new type of bolus that leverages most of the fields from `Normal` bolus. The subType associated to this type of bolus is `biphasic`.
+We add the following fields:
+- `bolusId`: unique ID provided by the client that is used to link the 2 parts of the bolus.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]: We introduce the eventId field to link deviceEvents objects, and here we use bolusId field to link bolus objects. 

I'm not sure we need to create two different fields. The drawbacks I see : 
 - we will have to create two indices in the future if we have to optimize queries to get linked objects
 - data clients (i.e. blip for now) will have two processes to link objects together

**Review by CedricPerotto — CHANGES_REQUESTED** [review]

File: docs/UpdatedData.md (hunk 14 above)
Comment (franck-fourel) [reply docs/UpdatedData.md:251]: yep, that's correct

File: docs/UpdatedData.md (hunk 15 above)
Comment (franck-fourel) [reply docs/UpdatedData.md:274]: correct

File: docs/UpdatedData.md (hunk 9 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: yep sounds better

File: docs/UpdatedData.md (hunk 16 above)
Comment (franck-fourel) [reply docs/UpdatedData.md]: Good point. 
let's use the same field for linking objects.

**Review by CedricPerotto — APPROVED** [review]

**Pull request merged**